File: providers/ollama.py
import logging
from typing import List, Dict

from ollama import Client as OllamaClient

from config.settings import settings
from providers.base import AIProvider

logger = logging.getLogger(__name__)


class OllamaProvider(AIProvider):
    """Ollama Cloud provider — uses the Ollama Python SDK with streaming."""

    def __init__(self):
        self._client = OllamaClient(
            host=settings.OLLAMA_HOST,
            headers={"Authorization": f"Bearer {settings.OLLAMA_API_KEY}"},
        )
        self._model = settings.OLLAMA_MODEL
        logger.info("OllamaProvider initialized (host=%s, model=%s)", settings.OLLAMA_HOST, self._model)

    def chat(self, messages: List[Dict[str, str]]) -> str:
        reply = ""
        for part in self._client.chat(
            self._model, messages=messages, stream=True
        ):
            content = part.get("message", {}).get("content", "")
            reply += content
        return reply

    def health_check(self) -> bool:
        try:
            # A minimal request to verify connectivity
            self._client.list()
            return True
        except Exception as e:
            logger.error("Ollama health check failed: %s", e)
            return False

File: src/providers/ollama.ts
import axios from "axios";
import { AIProvider, ImageDescription } from "./base";
import { Settings } from "../config/settings";
import { ChatMessage, toDict } from "../storage/models";

export class OllamaProvider implements AIProvider {
    private host: string;
    private model: string;
    private headers: Record<string, string>;

    constructor() {
        this.host = Settings.OLLAMA_HOST;
        this.model = Settings.OLLAMA_MODEL;
        this.headers = {
            "Authorization": `Bearer ${Settings.OLLAMA_API_KEY}`,
            "Content-Type": "application/json"
        };
        console.log(`OllamaProvider initialized (host=${this.host}, model=${this.model})`);
    }

    async chat(messages: ChatMessage[]): Promise<string> {
        const payload = {
            model: this.model,
            messages: messages.map(toDict),
            stream: false
        };

        const response = await axios.post(`${this.host}/api/chat`, payload, {
            headers: this.headers
        });

        return response.data.message.content;
    }

    async describeImage(imageBuffer: Buffer, mimeType: string): Promise<ImageDescription> {
        // Convert the image buffer to base64
        const base64Image = imageBuffer.toString("base64");

        // The data URI format for the image
        const dataUri = `data:${mimeType};base64,${base64Image}`;

        // Prompt specifically for image description and bank transaction detection
        const prompt = `You are an image analysis assistant. Analyze this image carefully.

Describe what you see in this image in detail. If this image shows a bank transaction receipt, bank transfer confirmation, money transfer slip, or any financial transaction proof, you MUST start your response with exactly "BANK_TRANSACTION:" followed by the description.

If it is NOT a bank transaction or financial receipt, just provide a clear description of what the image contains.

Be concise but thorough in your description.`;

        const payload = {
            model: this.model,
            messages: [
                {
                    role: "user",
                    content: prompt,
                    images: [dataUri]
                }
            ],
            stream: false
        };

        const response = await axios.post(`${this.host}/api/chat`, payload, {
            headers: this.headers
        });

        const content = response.data.message.content;
        const isBankTransaction = content.startsWith("BANK_TRANSACTION:");

        // Clean the description by removing the BANK_TRANSACTION prefix if present
        const description = isBankTransaction
            ? content.replace("BANK_TRANSACTION:", "").trim()
            : content.trim();

        return {
            description,
            isBankTransaction
        };
    }

    async healthCheck(): Promise<boolean> {
        try {
            await axios.get(`${this.host}/api/tags`, { headers: this.headers });
            return true;
        } catch (e) {
            console.error("Ollama health check failed:", e);
            return false;
        }
    }
}